lua/utils: use `index2adr()` instead of index arithmetic

Gumix opened the pull request:
We have a code that performs calculations to obtain the address from the
given index, the address is then passed to `cdataV()` to get cdata value from
the stack. But this doesn't work for pseudo-indexes (e.g. upvalue indexes).
This patch brings `index2adr()` from `luajit/src/lj_api.c`, which accepts all
kinds of indexes, so that the calculations are no longer needed. Also the
helper function `luaL_tocpointer()` is introduced.

Closes #8249

igormunkin commented on src/lua/utils.c:
```diff
@@ -53,6 +53,37 @@ uint32_t CTID_UUID;
 
+/** A copy of index2adr() from luajit/src/lj_api.c */
+static TValue *
+luaL_index2adr(lua_State *L, int idx)
```
Generally, we try to avoid using `luaL` prefix in function naming (see #4577 for more info), but this still respects our current guidelines and this function is not exported, so can be freely renamed later. Feel free to ignore.

Gumix replied:
Renamed to `index2adr()`.

igormunkin commented on src/lua/utils.c:
```diff
@@ -554,17 +576,13 @@ luaT_tolstring(lua_State *L, int idx, size_t *len)
 	return lua_tolstring(L, -1, len);
 }
 
-/* Based on ffi_meta___call() from luajit/src/lib_ffi.c. */
```
Typo: According to our [guidelines](https://www.tarantool.io/en/doc/latest/dev_guide/c_style_guide/#chapter-8-commenting) dots are required in comments. Same for the comment above.

Gumix replied:
Fixed.

igormunkin reviewed (changes requested):
@Gumix, thanks for the patch! Besides the comments below, I also grepped all sources to find all `L->base` occurrences. There is one in `src/lua/utils.h` and two in `src/lua/serializer.c`.

Considering these usages, I'd rather suggest to introduce the helper to obtain cdata payload.

Gumix commented:
> @Gumix, thanks for the patch! Besides the comments below, I also grepped all sources to find all `L->base` occurrences. There is one in `src/lua/utils.h` and two in `src/lua/serializer.c`.
> 
> Considering these usages, I'd rather suggest to introduce the helper to obtain cdata payload.

Added `get_cdata()`.

> P.S. I also think, that changelog is required, since you change the semantics of the public interfaces.

Added changelog and testcase.

igormunkin commented on src/lua/utils.c:
```diff
@@ -54,6 +54,43 @@ uint32_t CTID_UUID;
+}
+
+GCcdata *
+get_cdata(struct lua_State *L, int idx)
```
OK, this is not quite the thing I meant in the previous review. I propose to implement a new function `luaL_tocpointer(lua_State *L, int idx, uint32_t *ctypeid)` with the signature similar to `lua_tolstring` and implementing `luaL_checkcdata` without actual check. I left an example below.

```c
void *
luaL_tocpointer(lua_State *L, int idx, uint32 *ctypeid) {
        TValue *o = index2adr(L, idx);
        if (!tviscdata(o)) {
                if (ctypeid != NULL) *ctypeid = 0;
                return NULL;
        }
        GCcdata *cd = cdataV(o);
        if (ctypeid != NULL) *ctypeid = cd->ctypeid;
        return cdataptr(cd);
}
```

IMHO, such function can be used widely in Tarantool to work with cdata without additional checks.

Gumix replied:
Done.

igormunkin reviewed (changes requested):
@Gumix, thanks for the changes! I've left some more comments below.

The main purpose of introducing `luaL_tocpointer` is avoid raw access to `L->base`. For other enhancements (if there is something you want to change) I suggest to open a new PR.

BTW, I'd rather split introducing of `index2adr` helper and `luaL_tocpointer` function into 2 separate commits. This is just recommendation, so feel free to ignore.

Gumix commented on src/lua/serializer.c:
```diff
@@ -447,9 +465,6 @@ int
 {
-	if (index < 0)
-		index = lua_gettop(L) + index + 1;
-
```
Reverted.

igormunkin reviewed (approved):
@Gumix, thanks for the fixes! LGTM.
